command/agent/prepared_query_endpoint.go

```diff
@@ -0,0 +1,190 @@
+package agent
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/consul/consul"
+	"github.com/hashicorp/consul/consul/structs"
+)
+
+const (
+	preparedQueryEndpoint      = "PreparedQuery"
+	preparedQueryExecuteSuffix = "/execute"
+)
+
+// preparedQueryCreateResponse is used to wrap the query ID.
+type preparedQueryCreateResponse struct {
+	ID string
+}
+
+// PreparedQueryGeneral handles all the general prepared query requests.
+func (s *HTTPServer) PreparedQueryGeneral(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	endpoint := s.agent.getEndpoint(preparedQueryEndpoint)
+	switch req.Method {
+	case "POST": // Create a new prepared query.
+		args := structs.PreparedQueryRequest{
+			Op: structs.PreparedQueryCreate,
+		}
+		s.parseDC(req, &args.Datacenter)
+		s.parseToken(req, &args.Token)
+		if req.ContentLength > 0 {
+			if err := decodeBody(req, &args.Query, nil); err != nil {
+				resp.WriteHeader(400)
+				resp.Write([]byte(fmt.Sprintf("Request decode failed: %v", err)))
+				return nil, nil
+			}
+		}
+
+		var reply string
+		if err := s.agent.RPC(endpoint+".Apply", &args, &reply); err != nil {
+			return nil, err
+		}
+		return preparedQueryCreateResponse{reply}, nil
+
+	case "GET": // List all the prepared queries.
+		var args structs.DCSpecificRequest
+		if done := s.parse(resp, req, &args.Datacenter, &args.QueryOptions); done {
+			return nil, nil
+		}
+
+		var reply structs.IndexedPreparedQueries
+		if err := s.agent.RPC(endpoint+".List", &args, &reply); err != nil {
+			return nil, err
+		}
+
+		// Use empty list instead of nil.
+		if reply.Queries == nil {
+			reply.Queries = make(structs.PreparedQueries, 0)
+		}
+		return reply.Queries, nil
+
+	default:
+		resp.WriteHeader(405)
+		return nil, nil
+	}
+}
+
+// parseLimit parses the optional limit parameter for a prepared query execution.
+func parseLimit(req *http.Request, limit *int) error {
+	*limit = 0
+	if arg := req.URL.Query().Get("limit"); arg != "" {
+		if i, err := strconv.Atoi(arg); err != nil {
+			return err
+		} else {
+			*limit = i
+		}
+	}
+	return nil
+}
+
+// PreparedQuerySpecifc handles all the prepared query requests specific to a
+// particular query.
+func (s *HTTPServer) PreparedQuerySpecific(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
```

Comment:
This method is huge! Can we maybe make this function into a router, and have like a `preparedQueryGet`, `preparedQueryUpdate`, etc? Same for the method above, though it isn't as bad.